### PR TITLE
3 1 updated

### PR DESCRIPTION
**Pull request by himanshumishra31**

### CODE REVIEW

File: app/models/spree/taxjar.rb [hunk 1]
```diff
@@ -176,14 +171,18 @@ def address_params
           to_country: tax_address_country_iso,
           to_zip: tax_address_zip,
           to_state: tax_address_state_abbr,
-          to_city: tax_address_city
+          to_city: tax_address_city,
+          from_country: 'US',
```
Comment (sawangupta92) [inline app/models/spree/taxjar.rb]: ?? what's this? this was hardcoded in my app. I think I asked not to use the last commit.

File: app/models/spree/taxjar.rb [hunk 2]
```diff
@@ -21,22 +21,18 @@ def create_refund_transaction_for_order
 
     def create_transaction_for_order
       SpreeTaxjar::Logger.log(__method__, {order: {id: @order.id, number: @order.number}}) if SpreeTaxjar::Logger.logger_enabled?
-      if has_nexus?
```
Comment (sawangupta92) [inline app/models/spree/taxjar.rb]: I think I asked not to use the last commit.

File: app/models/spree/taxjar.rb [hunk 3]
```diff
@@ -21,22 +21,18 @@ def create_refund_transaction_for_order
...
+      api_params = transaction_parameters
+      SpreeTaxjar::Logger.log(__method__, {order: {id: @order.id, number: @order.number}, api_params: api_params}) if SpreeTaxjar::Logger.logger_enabled?
+      api_response = @client.create_order(api_params)
+      SpreeTaxjar::Logger.log(__method__, {order: {id: @order.id, number: @order.number}, api_response: api_response}) if SpreeTaxjar::Logger.logger_enabled?
+      api_response
     end
 
     def delete_transaction_for_order
       SpreeTaxjar::Logger.log(__method__, {order: {id: @order.id, number: @order.number}}) if SpreeTaxjar::Logger.logger_enabled?
-      if has_nexus?
```
Comment (sawangupta92) [inline app/models/spree/taxjar.rb]: We need the nexus

File: app/models/spree/calculator/taxjar_calculator.rb [hunk 4]
```diff
@@ -94,9 +94,9 @@ def cache_response(taxjar_response, order, address, item = nil)
 
       def cache_key(order, item, address)
         if item.is_a?(Spree::LineItem)
-          ['Spree::LineItem', order.id, item.id, address.state.id, address.zipcode, item.amount, :amount_to_collect]
+          ['Spree::LineItem', order.id, item.id, address.state.id, address.zipcode, item.taxable_amount, :amount_to_collect]
         else
-          ['Spree::Shipment', order.id, item.id, address.state.id, address.zipcode, item.cost, :amount_to_collect]
+          ['Spree::Shipment', order.id, item.id, address.state.id, address.zipcode, item.cost, item.adjustments.where.not(source_type: "Spree::TaxRate").pluck(:amount).sum.to_f, :amount_to_collect]
```
Comment (tanmay3011) [inline app/models/spree/calculator/taxjar_calculator.rb]: @himanshumishra31 Please avoid firing a query here and use already loaded objects to find the sum

Comment (tanmay3011) [inline app/models/spree/calculator/taxjar_calculator.rb]: ```address.state_id```

File: app/models/spree/taxjar.rb [hunk 5]
```diff
@@ -108,7 +108,7 @@ def tax_address
       def tax_params
         {
           amount: @order.item_total,
-          shipping: @order.shipment_total,
+          shipping: @order.shipment_total + @order.shipment_adjustments.where.not(source_type: "Spree::TaxRate").pluck(:amount).sum,
```
Comment (tanmay3011) [inline app/models/spree/taxjar.rb]: :ditto:

File: app/models/spree/taxjar.rb [hunk 6]
```diff
@@ -164,8 +163,8 @@ def transaction_parameters
         address_params.merge({
           transaction_id: @order.number,
           transaction_date: @order.completed_at.as_json,
-          amount: @order.item_total + @order.shipment_total,
-          shipping: @order.shipment_total,
+          amount: @order.total - @order.tax_total,
+          shipping: @order.shipment_total + @order.shipment_adjustments.where.not(source_type: "Spree::TaxRate").pluck(:amount).sum,
```
Comment (tanmay3011) [inline app/models/spree/taxjar.rb]: :ditto:

File: app/models/spree/calculator/taxjar_calculator.rb (hunk 4 above)
Comment (tanmay3011) [reply app/models/spree/calculator/taxjar_calculator.rb]: Spree::TaxRate.name or Spree::TaxRate.to_s

File: app/models/spree/taxjar.rb [hunk 7]
```diff
@@ -193,13 +192,19 @@ def line_item_params
...
         end
       end
 
+      def discount_weightage(item)
+        return 0 if @order.item_total.zero?
+        weightage = @order.adjustments.sum(:amount) / (@order.item_total)
+        - weightage * (item.taxable_amount / item.quantity)
+      end
+
   end
```
Comment (tanmay3011) [inline app/models/spree/taxjar.rb]: Avoid item.taxable_amount / item.quantity twice line 195 and 206

File: app/models/spree/taxjar.rb [hunk 8]
```diff
@@ -183,7 +182,7 @@ def address_params
       def shipment_tax_params
         address_params.merge({
           amount: 0,
-          shipping: @shipment.cost
+          shipping: @shipment.cost + @shipment.adjustments.where.not(source_type: "Spree::TaxRate").pluck(:amount).sum.to_f
```
Comment (tanmay3011) [inline app/models/spree/taxjar.rb]: :ditto:

**Pull request merged**